### ISSUE TITLE
Null vs absent fields

### ISSUE THREAD

**Issue by dbrgn** [coredump-ch/spaceapi#15]
Sometimes there's a difference between a null field and an absent field.

http://spaceapi.net/documentation

As an example, the `sensors.people_now_present.names` array may be absent, but if present it should contain at least 1 entry. (I think that's a bug though, should probably be reported to be fixed in a future API version.)

Also, `state.open` (which is stated to have the `Array` type, probably another bug) can be `true` (open), `false` (closed), `null` (we don't know, e.g. when the presence counter is down) and absent (when we don't tell).

In the end, we could probably get some of the details fixed in the next API version and use `null` values everywhere. But maybe there are still other places where null is not the same as the absence of the field.

Any ideas / opinions? @rnestler?

**Comment by dbrgn** [comment]
Reported both issues:

- https://github.com/SpaceApi/spaceapiapp_documentation/commit/3c322db4dacdea1b7f5bf8d8e0306cd48c186b42#commitcomment-11054369
- https://github.com/SpaceApi/SchemaValidator/issues/7

**Comment by rnestler** [comment]
I think for optional fields the ``Option<T>`` is the most natural thing.

**Comment by dbrgn** [comment]
Depends on the implementation. Should it? Probably, yes.

Some implementations might think "if there's a field there must be a value", but then the implementations are broken.

**Comment by slopjong** [comment]
Quickly reviewed the [specs](http://spaceapi.net/specs/0.13) and indeed, the documentation isn't consistent with the defined types.

```
"state":{
      ...
      "properties":{
        "open":{
          ...
          "type":[
            "boolean",
            "null"
          ]
        }
}
```

Basically yes, it looks like a bug in the documentation generator.

**Comment by dbrgn** [comment]
@slopjong Regarding the API in general, can we assume that a missing field should semantically be the same as a field set to `null`?

**Comment by slopjong** [comment]
The validator won't accept them in general.

**Comment by dbrgn** [comment]
Is that an API specification issue or a validator issue?

**Comment by slopjong** [comment]
It's more like a JSON schema *feature*. If a non-required field is present it must conform to the specified type. And validators implementing the the JSON schema guidelines will all behave like that.

**Comment by dbrgn** [comment]
Alright, then we have to find a better solution than setting the fields to `null` :)

**Comment by dbrgn** [comment]
(@slopjong: Lists should still be allowed to be empty though.)

**Comment by dbrgn** [comment]
According to today's discussion we'll simply [implement the ToJson trait](http://doc.rust-lang.org/rustc-serialize/rustc_serialize/json/index.html#using-the-tojson-trait) for all structs containing custom types.

**Comment by dbrgn** [comment]
Fixed by PR #21.
